core/src/main/java/org/apache/druid/query/QueryException.java

```diff
@@ -30,12 +29,129 @@
+ * occurred.
+ * 2. When these objects get serialized across the wire the recipient deserializes a QueryException. The client is
+ * never expected, and fundamentally is not allowed to, ever deserialize a child class of QueryException.
+ * <p>
+ * For this reason, QueryException must contain all potential state that any of its child classes could ever want to
+ * push across the wire.  Additionally, any catch clauses expecting one of the child Exceptions must know that it is
+ * running inside of code where the exception has not traveled across the wire.  If there is a chance that the
+ * exception could have been serialized across the wire, the code must catch a QueryException and check the errorCode
+ * instead.
+ * <p>
+ * As a corollary, adding new state or adjusting the logic of this class must always be done in a backwards-compatible
+ * fashion across all child classes of QueryException.
+ * <p>
+ * If there is any need to do different logic based on the type of error that has happened, the only reliable method
+ * of discerning the type of the error is to look at the errorCode String.  Because these Strings are considered part
+ * of the API, they are not allowed to change and must maintain their same semantics.  The known errorCode Strings
+ * are pulled together as public static fields on this class in order to make it more clear what options exist.
+ * <p>
  * QueryResource and SqlResource are expected to emit the JSON form of this object when errors happen.
  */
 public class QueryException extends RuntimeException implements SanitizableException
 {
+  /**
+   * Error codes
+   */
+  public static final String JSON_PARSE_ERROR_CODE = "Json parse failed";
+  public static final String BAD_QUERY_CONTEXT_ERROR_CODE = "Query context parse failed";
+  public static final String QUERY_CAPACITY_EXCEEDED_ERROR_CODE = "Query capacity exceeded";
+  public static final String QUERY_INTERRUPTED_ERROR_CODE = "Query interrupted";
+  // Note: the proper spelling is with a single "l", but the version with
+  // two "l"s is documented, we can't change the text of the message.
+  public static final String QUERY_CANCELED_ERROR_CODE = "Query cancelled";
+  public static final String UNAUTHORIZED_ERROR_CODE = "Unauthorized request";
+  public static final String UNSUPPORTED_OPERATION_ERROR_CODE = "Unsupported operation";
+  public static final String TRUNCATED_RESPONSE_CONTEXT_ERROR_CODE = "Truncated response context";
+  public static final String UNKNOWN_EXCEPTION_ERROR_CODE = "Unknown exception";
+  public static final String QUERY_TIMEOUT_ERROR_CODE = "Query timeout";
+  public static final String QUERY_UNSUPPORTED_ERROR_CODE = "Unsupported query";
+  public static final String RESOURCE_LIMIT_EXCEEDED_ERROR_CODE = "Resource limit exceeded";
+  public static final String SQL_PARSE_FAILED_ERROR_CODE = "SQL parse failed";
+  public static final String PLAN_VALIDATION_FAILED_ERROR_CODE = "Plan validation failed";
```

Comment:
These are all quite lame messages. If I was a user, and was told "Plan validation failed", I'd have no idea where to start to set things right. Is this a user error? If so, what did I do wrong? Or, did the system run out of resources? Is there a bug in Druid: the plan should have worked, but didn't.

We really gotta put some effort into improving our messages on top of just making the lousy messages into nice constants.

Reply:
I agree, that said, these are not error messages, these are error codes.  These are the current code's equivalent to a numerical error code.  The change that I did is not trying to make them into constants, instead, I realized that they were being used equivalent to a numerical code and so I decided that I wanted them all in one place so that we can actually see which "codes" we are using for errors.  I.e. I just pulled them together from around the code base to this single location.